redisson/src/main/java/org/redisson/RedissonSpinLock.java

```diff
@@ -0,0 +1,481 @@
+/**
+ * Copyright (c) 2013-2020 Nikita Koksharov
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.redisson;
+
+import org.redisson.api.RFuture;
+import org.redisson.client.RedisException;
+import org.redisson.client.codec.LongCodec;
+import org.redisson.client.protocol.RedisCommands;
+import org.redisson.client.protocol.RedisStrictCommand;
+import org.redisson.command.CommandAsyncExecutor;
+import org.redisson.misc.RPromise;
+import org.redisson.misc.RedissonPromise;
+
+import java.util.Collections;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Distributed implementation of {@link java.util.concurrent.locks.Lock}
+ * Implements reentrant lock.<br>
+ * Lock will be removed automatically if client disconnects.
+ * This lock implementation doesn't use pub/sub mechanism. It can be used in large Redis clusters despite current naive
+ * pub/sub implementation.
+ * <p>
+ * Implements a <b>non-fair</b> locking so doesn't guarantees an acquire order.
+ *
+ * @author Danila Varatyntsev
+ */
+public class RedissonSpinLock extends RedissonBaseLock {
+
+    public interface BackOffOptions {
+        BackOffPolicy create();
+    }
+
+    public interface BackOffPolicy {
+
+        long getNextSleepPeriod();
+    }
+
+    public static class ExponentialBackOffOptions implements BackOffOptions {
```

Comment:
These classes should be in implementation class. You need to introduce LockOptions object similar to MapOptions and put them there.